(RE-4221) Allow building and shipping of only tar/gem artifacts

melissa opened the pull request:
With the shift to delivering our software via the puppet-agent package,
we want to discontinue shipping conventional packages. However, we do
want to continue shipping the tar and gem artifacts.

This commit lets us maintain a list of LTS build targets for community
members who are using this automation to build rpm or deb packages while
not building rpm/deb packages when we run `rake pl:jenkins:uber_build`.
It also lets us easily not ship any deb/rpm packages. The important this
is that `pl:deb` and `pl:mock` still work for building deb/rpm packages
if that is wanted. If the setting needs to be overridden, like for a
jenkins uber_build, we can do that by setting `RPM=TRUE` and/or
`DEB=TRUE` as environment variables when that rake task is invoked.

melissa commented:
http://jenkins-release.delivery.puppetlabs.net/job/hiera-packaging-2015-03-25-15-04-59-c824062101744e6fac5b2ea9372dc0a075565bdb/ was built with 

```diff
commit c824062101744e6fac5b2ea9372dc0a075565bdb
Author: Melissa Stone <melissa@puppetlabs.com>
Date:   Wed Mar 25 14:51:13 2015 -0700

    (maint) Only build LTS rpm/deb targets

diff --git a/ext/build_defaults.yaml b/ext/build_defaults.yaml
index bc0d56c..423ad1c 100644
--- a/ext/build_defaults.yaml
+++ b/ext/build_defaults.yaml
@@ -2,16 +2,21 @@
 packaging_url: 'git://github.com/puppetlabs/packaging.git --branch=master'
 packaging_repo: 'packaging'
 default_cow: 'base-squeeze-i386.cow'
-cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-stable-i386.cow base-testing-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
+cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
 pbuild_conf: '/etc/pbuilderrc'
 packager: 'puppetlabs'
 gpg_name: 'info@puppetlabs.com'
 gpg_key: '4BD6EC30'
 sign_tar: FALSE
 # a space separated list of mock configs
-final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64 pl-fedora-19-i386 pl-fedora-20-i386'
+final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64'
 build_gem: TRUE
-build_dmg: TRUE
+build_dmg: FALSE
+
+# Do not build rpm/deb packages during uber_build
+# Do not ship rpm/deb packages during uber_ship
+build_rpm: FALSE
+build_deb: FALSE
 yum_host: 'yum.puppetlabs.com'
 yum_repo_path: '/opt/repository/yum/'
 apt_host: 'apt.puppetlabs.com'
```


and http://jenkins-release.delivery.puppetlabs.net/job/hiera-packaging-2015-03-25-15-05-45-f6d9b00fb7c6f1b188678f9e30d8ddd0208745ef/ was built with 
```diff
diff --git a/ext/build_defaults.yaml b/ext/build_defaults.yaml
index bc0d56c..15ba294 100644
--- a/ext/build_defaults.yaml
+++ b/ext/build_defaults.yaml
@@ -2,16 +2,16 @@
 packaging_url: 'git://github.com/puppetlabs/packaging.git --branch=master'
 packaging_repo: 'packaging'
 default_cow: 'base-squeeze-i386.cow'
-cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-stable-i386.cow base-testing-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
+cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
 pbuild_conf: '/etc/pbuilderrc'
 packager: 'puppetlabs'
 gpg_name: 'info@puppetlabs.com'
 gpg_key: '4BD6EC30'
 sign_tar: FALSE
 # a space separated list of mock configs
-final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64 pl-fedora-19-i386 pl-fedora-20-i386'
+final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64'
 build_gem: TRUE
-build_dmg: TRUE
+build_dmg: FALSE
 yum_host: 'yum.puppetlabs.com'
 yum_repo_path: '/opt/repository/yum/'
 apt_host: 'apt.puppetlabs.com'
```

MosesMendoza commented:
First, just to make sure I understand - by default, this does not change any behavior? A project has to set build_deb=false and/or build_rpm=false to not build these packages with the uber_build?

Second, I'd suggest that the names be changed for the variables to something that signifies that they only affect the outcome of the uber_build, e.b., uber_build_rpm or some such thing. Otherwise its inconsistent with the other build_* variables. For example, If build_dmg is false, you can't build a dmg at all - the task isn't even loaded. Same with build_gem. It would be odd to specify build_rpm=false and still have all the rpm tasks show up and execute like before.

melissa commented:
@MosesMendoza thanks for looking this over!

By default, this should not change any behavior by default. If `uber_build_deb` or `uber_build_rpm` are not set, they will default to true, and deb/rpm packages will be built as they would normally. If a project does not want to build deb or rpm packages, they have to explicitly set `uber_build_rpm: FALSE` the the build_defaults file, or set `UBER_BUILD_RPM=FALSE` as an environment variable during the rake invocation.

I've updated the variable name to conform with what you said. I've also updated the environmental var overrides to be `UBER_BUILD_DEB` and `UBER_BUILD_RPM`

puppetcla commented:
CLA signed by all contributors.

melissa commented:
Talking this over with @haus we should be able to override the contents of `Pkg::Config.cows` and `Pkg::Config.final_mocks` in build-data (the release branch). This way, we can maintain a list community members can use if they are using the packaging repo to build packages of our projects. This means we get to remove the `uber_build_deb` and `uber_build_rpm` vars, and just check if our cow/mock list is empty.

However, one problem I do see with this is if we are in fact still building rpm/deb packages for certain projects (like the mco plugins). This would require maintaining a cow/mock list for those projects in build-data to override the release team override. Although that isn't a difficult thing to maintain, it does require some extra attention to remember to update that additional repo, and which projects require updates/

melissa commented:
So, @haus, thinking about this more, I do prefer your suggestion with updating this in build-data rather than here, but I think that, for our needs, this is going to be the best way to deal with it. There are too many repos that we'd need to override in order to continue building packages for those projects that we still need to build packages for. Like for the puppet 3.x series, puppetdb 2.x, mco 2.8, etc. Once we're no longer building/shipping packages for those series, I'd vote we revert this commit, and implement your suggestion. Thoughts?

melissa commented:
@haus did we come to a consensus about how to deal with this?

melissa commented:
@haus this would be super awesome to merge, since we can't make your suggestion work across different branches

haus commented:
@melissa I'll respond this morning.

haus commented:
@melissa i was thinking it might be reasonable to have mocks and cows empty, and have a commented list of LTS targets above in the yaml file. don't know how crazy that is.

haus commented:
By the way, my trying to think of alternatives is mainly trying to reduce complexity here and not add new params. I don't think there's anything wrong with the PR as is.

melissa commented:
@haus yup, definitely reasonable. I'd be in favor of that. That way, we can pair down this to just check if cow/mock targets are empty

melissa commented:
I mostly just want to avoid surprises for users

haus commented:
absolutely.

melissa commented:
@haus like this? I need to squash and test this iteration, also

haus commented on templates/packaging.xml.erb:
```diff
@@ -107,8 +111,12 @@ return labelMap.get(binding.getVariables().get(&quot;command&quot;));</groovyScr
+        <% if not Pkg::Config.cows.empty? %>
+          <% Pkg::Config.cows.split(' ').each do |cow| %><string>pl_deb COW=<%= cow %></string><% end %>
+        <% end %>
+        <% if now Pkg::Config.final_mocks.empty? %>
```
now => not

haus commented:
@melissa with simple (single clause) conditionals, i find `unless` more readable than `if not`. Most (all?) of the conditionals in the PR fall under that category. Using unless is also more consistent with the rest of the packaging repo.

haus commented:
@melissa travis and rubocop seem to agree with me :)

melissa commented:
build http://jenkins-release.delivery.puppetlabs.net/job/puppet-packaging-2015-04-08-15-09-11-6b997288a72792b773498bf7fa00fd2f6c3ad8b3/ with the following:

```diff
diff --git a/ext/build_defaults.yaml b/ext/build_defaults.yaml
index 5bbf63d..01e6c78 100644
--- a/ext/build_defaults.yaml
+++ b/ext/build_defaults.yaml
@@ -1,47 +1,21 @@
 ---
 packaging_url: 'git://github.com/puppetlabs/packaging.git --branch=master'
 packaging_repo: 'packaging'
-default_cow: 'base-squeeze-i386.cow'
-cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-stable-i386.cow base-testing-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
-pbuild_conf: '/etc/pbuilderrc'
 packager: 'puppetlabs'
-gpg_name: 'info@puppetlabs.com'
 gpg_key: '4BD6EC30'
+
+# These are the build targets used by the packaging repo. Uncomment to allow use.
+final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64'
+#default_cow: 'base-squeeze-i386.cow'
+cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
+pbuild_conf: '/etc/pbuilderrc'
+
+build_gem: TRUE
+build_dmg: FALSE
 sign_tar: FALSE
-# a space separated list of mock configs
-final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64 pl-fedora-19-i386 pl-fedora-20-i386'
 yum_host: 'yum.puppetlabs.com'
 yum_repo_path: '/opt/repository/yum/'
-build_gem: TRUE
-build_dmg: TRUE
-msi_name: 'puppet-agent'
-build_msi:
-  puppet_for_the_win:
-    ref: '52e81967253924428bb0b1e7e1a11e8678d68378'
-    repo: 'git://github.com/puppetlabs/puppet_for_the_win.git'
-  facter:
-    ref: 'refs/tags/2.4.0'
-    repo: 'git://github.com/puppetlabs/facter.git'
-  cfacter:
-    archive:
-      x86: 'cfacter-0.3.0-x86.zip'
-      x64: 'cfacter-0.3.0-x64.zip'
-    path: 'http://builds.puppetlabs.lan/cfacter/0.3.0/artifacts/windows'
-  hiera:
-    ref: 'refs/tags/2.0.0'
-    repo: 'git://github.com/puppetlabs/hiera.git'
-  mcollective:
-    ref: 'refs/tags/2.7.0'
-    repo: 'git://github.com/puppetlabs/marionette-collective.git'
-  sys:
-    ref:
-      x86: '8db9d84da9950760144b5dfcd807213eecee4842'
-      x64: '12030f11e9bb2f085c68108bff34be6956b25df9'
-    repo: 'git://github.com/puppetlabs/puppet-win32-ruby.git'
 apt_host: 'apt.puppetlabs.com'
 apt_repo_url: 'http://apt.puppetlabs.com'
 apt_repo_path: '/opt/repository/incoming'
-ips_repo: '/var/pkgrepo'
-ips_store: '/opt/repository'
-ips_host: 'solaris-11-ips-repo.acctest.dc1.puppetlabs.net'
 tar_host: 'downloads.puppetlabs.com'
```

and build http://jenkins-release.delivery.puppetlabs.net/job/puppet-packaging-2015-04-08-15-15-08-d60bb54d52f0f111decb0217fc7c3c84ac835dee/ (which failed because of unsatisfied dependencies) with
```diff
diff --git a/ext/build_defaults.yaml b/ext/build_defaults.yaml
index 5bbf63d..053ae43 100644
--- a/ext/build_defaults.yaml
+++ b/ext/build_defaults.yaml
@@ -1,47 +1,23 @@
 ---
 packaging_url: 'git://github.com/puppetlabs/packaging.git --branch=master'
 packaging_repo: 'packaging'
-default_cow: 'base-squeeze-i386.cow'
-cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-stable-i386.cow base-testing-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
-pbuild_conf: '/etc/pbuilderrc'
 packager: 'puppetlabs'
-gpg_name: 'info@puppetlabs.com'
 gpg_key: '4BD6EC30'
+
+# These are the build targets used by the packaging repo. Uncomment to allow use.
+#final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64'
+final_mocks:
+#default_cow: 'base-squeeze-i386.cow'
+#cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
+cows:
+pbuild_conf: '/etc/pbuilderrc'
+
+build_gem: TRUE
+build_dmg: FALSE
 sign_tar: FALSE
-# a space separated list of mock configs
-final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64 pl-fedora-19-i386 pl-fedora-20-i386'
 yum_host: 'yum.puppetlabs.com'
 yum_repo_path: '/opt/repository/yum/'
-build_gem: TRUE
-build_dmg: TRUE
-msi_name: 'puppet-agent'
-build_msi:
-  puppet_for_the_win:
-    ref: '52e81967253924428bb0b1e7e1a11e8678d68378'
-    repo: 'git://github.com/puppetlabs/puppet_for_the_win.git'
-  facter:
-    ref: 'refs/tags/2.4.0'
-    repo: 'git://github.com/puppetlabs/facter.git'
-  cfacter:
-    archive:
-      x86: 'cfacter-0.3.0-x86.zip'
-      x64: 'cfacter-0.3.0-x64.zip'
-    path: 'http://builds.puppetlabs.lan/cfacter/0.3.0/artifacts/windows'
-  hiera:
-    ref: 'refs/tags/2.0.0'
-    repo: 'git://github.com/puppetlabs/hiera.git'
-  mcollective:
-    ref: 'refs/tags/2.7.0'
-    repo: 'git://github.com/puppetlabs/marionette-collective.git'
-  sys:
-    ref:
-      x86: '8db9d84da9950760144b5dfcd807213eecee4842'
-      x64: '12030f11e9bb2f085c68108bff34be6956b25df9'
-    repo: 'git://github.com/puppetlabs/puppet-win32-ruby.git'
 apt_host: 'apt.puppetlabs.com'
 apt_repo_url: 'http://apt.puppetlabs.com'
 apt_repo_path: '/opt/repository/incoming'
-ips_repo: '/var/pkgrepo'
-ips_store: '/opt/repository'
-ips_host: 'solaris-11-ips-repo.acctest.dc1.puppetlabs.net'
```

melissa commented:
@haus updated!!

haus commented on tasks/ship.rake:
```diff
@@ -129,8 +129,8 @@ namespace :pl do
       Rake::Task["pl:ship_gem"].invoke if Pkg::Config.build_gem
-      Rake::Task["pl:ship_rpms"].invoke
-      Rake::Task["pl:ship_debs"].invoke
+      Rake::Task["pl:ship_rpms"].invoke unless Pkg::Config.final_mocks.nil?
```
I think these should be `.invoke if Pkg::Config.final_mocks` instead of unless .nil?

haus commented:
In general, using `if thing` should be preferred over `unless thing.nil?`.

haus commented:
@melissa this looks good to me. have you tested builds with it?

melissa commented:
@haus yes, I tested with the linked pull request changing `ext/build_defaults.yaml` in puppet.

Tested with the following combinations:
```
cows:
final_mocks:
```
```
cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64'
```
```
#cows: 'base-lucid-i386.cow base-precise-i386.cow base-squeeze-i386.cow base-trusty-i386.cow base-wheezy-i386.cow'
#final_mocks: 'pl-el-5-i386 pl-el-6-i386 pl-el-7-x86_64'
```

haus commented:
cool. i'm :+1:
